core/src/banking_stage.rs

```diff
@@ -1883,6 +1885,61 @@ impl BankingStage {
         }
     }
 
+    /// This function filters extraneous consensus transactions. Given that votes are not
+    /// incremental, it suffices to only process the latest vote for each validator.
+    fn filter_extraneous_votes(
+        transactions: &[SanitizedTransaction],
+        transaction_to_packet_indexes: &[usize],
+    ) -> Vec<usize> {
+        // Store the latest vote for each validator
+        let mut vote_ixs_by_pubkey: HashMap<Pubkey, (Slot, usize)> = HashMap::new();
+        let non_votes = transactions
+            .iter()
+            .enumerate()
+            .filter_map(|(i, tx)| {
+                if tx.message().instructions().len() == 1 {
+                    let (program_pubkey, instruction) =
+                        tx.message().program_instructions_iter().next().unwrap();
+                    if program_pubkey == &solana_vote_program::id() {
+                        if let Ok(vote_instruction) =
+                            limited_deserialize::<VoteInstruction>(&instruction.data)
+                        {
+                            match vote_instruction {
+                                VoteInstruction::UpdateVoteState(vote_state_update)
+                                | VoteInstruction::UpdateVoteStateSwitch(vote_state_update, _) => {
+                                    // Only filter the full votes
+                                    let &vote_account_key = tx
+                                        .message()
+                                        .account_keys()
+                                        .get(0)
+                                        .expect("Vote account pubkey is missing");
+                                    let slot = vote_state_update.last_voted_slot().unwrap();
+                                    let cur_latest_slot = vote_ixs_by_pubkey
+                                        .get(&vote_account_key)
+                                        .map(|&(slot, _)| slot)
+                                        .unwrap_or_default();
+                                    if slot > cur_latest_slot {
+                                        // Only  keep the latest vote by slot
+                                        vote_ixs_by_pubkey.insert(vote_account_key, (slot, i));
+                                    }
+                                    return None;
+                                }
+                                _ => (),
+                            };
+                        }
+                    }
+                }
+                Some(i)
+            })
+            .collect_vec();
+        vote_ixs_by_pubkey
+            .into_iter()
+            .map(|(_, (_, i))| i)
+            .chain(non_votes.into_iter())
+            .map(|i| transaction_to_packet_indexes[i])
+            .collect_vec()
+    }
+
```

Comment:
This logic is correct I think, but perhaps in the wrong place.

So the issue here is that where `filter_extraneous_votes` is currently called, it will only analyze across the 128 packets that are passed to `process_packets_transactions` from the `MinMaxHeap` of pending work here:  https://github.com/solana-labs/solana/blob/980d056ce196ec21acdb983afe0ecfd6567aabb4/core/src/banking_stage.rs#L665-L689

I think what we want instead, is to only keep the latest vote for each validator across the entire buffer. 

I think to do this properly is quite involved:

1. I think for voting threads, banking stage should ideally use a different data structure to track the unprocessed work. I would imagine it could just be a `latest_votes: HashMap<Pubkey, Packet>` where we keep track of the latest packet.

2. We update `latest_votes` when we get a new vote transaction for a particular validator. It would be awesome here if the vote transaction format supported letting us quickly parse out what the latest vote here was, kind of like how we can quickly determine if a vote is a simple vote transaction:  https://github.com/solana-labs/solana/blame/980d056ce196ec21acdb983afe0ecfd6567aabb4/perf/src/sigverify.rs#L423-L429. That way we can quickly figure out when adding to the pending buffer whether this vote should be added/dropped

3.  With this `latest_votes` structure, we can then also always prioritize the highest staked validators, and drop votes from non-staked validators

Reply:
That being said, a easy first win would be to optimize the gossip vote sender to just send the latest vote per validator here: https://github.com/solana-labs/solana/blob/980d056ce196ec21acdb983afe0ecfd6567aabb4/core/src/verified_vote_packets.rs#L81-L128 😃

Reply:
I'd second the thought of putting vote into its own processing within banking stage. It'd be a project by itself, but I can see benefits.

Reply:
> We update latest_votes when we get a new vote transaction for a particular validator. It would be awesome here if the vote transaction format supported letting us quickly parse out what the latest vote here was, kind of like how we can quickly determine if a vote is a simple vote transaction: https://github.com/solana-labs/solana/blame/980d056ce196ec21acdb983afe0ecfd6567aabb4/perf/src/sigverify.rs#L423-L429. That way we can quickly figure out when adding to the pending buffer whether this vote should be added/dropped

Should it be filtered out in sigverify before it even makes it to banking stage?

Reply:
I can use a thread safe data structure to keep track of latest slot voted per validator and have the banking stage threads drop extraneous votes, however it seems like it would be nice if we could drop them earlier?

Problem is we need to be able to find the slot before we deserialize the transaction...